### ISSUE TITLE
Auto dispose after first listener

### ISSUE THREAD

**Issue by talent-apps** [rrousselGit/riverpod#1329]
Riverpod version `2.0.0-dev.4` adds a `disposeDelay` feature to all `autoDispose` providers and to `ProviderContainer`/`ProviderScope`. This configures the amount of time before a provider is disposed when it is not listened.

To avoid timing issues, it would be nice to add a "auto dispose after the provider is listened to once" mode, which make the provider auto disposable only after it is listened to once.

Example:
- Some function on screen A initalizes a provider with some initial value and then navigates to screen B.
- Only screen B watches the provider.
- Once screen B is disposed, the provider should be disposed.

**Comment by rrousselGit** [comment]
> * Only screen B watches the provider.

I would suggest have A listen to the provider until navigating to B

**Comment by talent-apps** [comment]
The nature of the use case is that A is not interested in this provider.
A classic example is a master/details scenario:
- Screen A presents a list of items
- Screen B allows to edit a single item
- The provider represents the item that is currently being edited.

This scenario is very common and the library doesn't seem to provide an easy way to handle it.

**Comment by rrousselGit** [comment]
Maybe it's worth expanding on what "initializing with some initial value" means and also explain what's the problem you're currently facing

**Comment by talent-apps** [comment]
The provider will be initalized with the item that was selected in the list.
Then the app will navigate to screen B where you can edit the properties of the item.

**Comment by rrousselGit** [comment]
> The provider will be initalized with the item that was selected in the list.

Right, but what exactly does this involve?

**Comment by talent-apps** [comment]
Code example:
```
void onItemSelected(BuildContext context, WidgetRef ref, Item selectedItem) {
  ref.read(editedItemProvider.notifier).setItem(selectedItem);
  Navigator.of(context).pushNamed(screenB);
}
```

```
class Item {
  final String name;
  const Item(this.name);

  Item copyWith(String? name) => Item(name ?? this.name);
}

final editedItemProvider = StateNotifierProvider.autoDispose<EditedItemNotifier, Item>((ref) {
  return EditedItemNotifier(ref: ref);
});

class EditedItemNotifier extends StateNotifier<Item?> {
  final Ref ref;

  EditedItemNotifier({required this.ref}) : super(null);

  void setItem(Item item) {
    state = item;
  }

  void updateName(String newName) {
    state = state?.copyWith(newName);
  }
}
```

**Comment by rrousselGit** [comment]
And why don't you want to set disposeDelay in this case?

**Comment by talent-apps** [comment]
It would work, but relying on a specific timeout is more error prune while relying on adding/removing a listener is exact.
With a timeout you alaso hold the state longer than you actually need it which is a disadvantage.
This scenario is very common in many apps and it would be nice to have this if it's a small feature to implement.

**Comment by robpot95** [comment]
I agree in this case, even tho disposeDelay help the issuw. Still uncertain how long it should hold the state. There is cases when I switch screen and then back to the old screen. I would like to dispose it immediately and not delay the dispose.

**Comment by danielkerwin** [comment]
I've been dealing with this same issue - my use-case is initializing a AutoDisposing StateNotifierProvider that is used to cache form values until submission - the initialization occurs when "master" form widget is loaded, but this widget does not care about the state of the provider; the detailed form widgets subscribe to different parts of the form state via the provider.

I cannot use auto-dispose unless the master watches the provider, causing rebuilds in the entire tree - having the auto-dispose occur only after at least one listen would solve my issue.

**Comment by rrousselGit** [comment]
Well ultimately you can do that yourself already

```
Provider((ref) {
  ref.onCancel(ref.invalidateSelf);
}
``

This effectively produces the same result as `autoDispose`, without the state getting destroyed during navigation

**Comment by rrousselGit** [comment]
While the proposal is interesting, I'm not sure it's ultimately a good idea.

I could see a few flaws:
- what if the provider is never listened?
- what if you have:

   ```
    final a = Provider.autoDispose(..., waitUntilListenedBeforeDispose);

    final b = Provider((ref) { 
       ref.watch(a);
       return Something();
    });

   // inside some widget:
   ref.read(b);
   Navigator.push();
  ```
  That'd still cause `a` to be disposed, since it's temporarily listened by `b`

**Comment by rrousselGit** [comment]
If we want to do something other than disposeDelay (which I personally think is good enough for the problem described), I'd personally prefer having a way to make `A` temporarily listen to the provider until `B` is poped

Like maybe:

```dart
Button(
  onTap: () async {
    final sub = ref.listen(provider, (prev, value) {});
    await Navigator.push(); 
    sub.close();
  },
)
```

**Comment by robpot95** [comment]
> While the proposal is interesting, I'm not sure it's ultimately a good idea.
> 
> I could see a few flaws:
> 
> * what if the provider is never listened?
> * what if you have:
>   ```
>    final a = Provider.autoDispose(..., waitUntilListenedBeforeDispose);
>   
>    final b = Provider((ref) { 
>       ref.watch(a);
>       return Something();
>    });
>   
>   // inside some widget:
>   ref.read(b);
>   Navigator.push();
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   That'd still cause `a` to be disposed, since it's temporarily listened by `b`

I agree with you, that this will cause a issue. But tbh when i set the provider state and navigate screen. First problem i'm bit unsure how long the disposeDelay should last. Maybe navigate screen takes a microsecond or couple seconds. Then i have the problem with that i would like to use the state in screen B that is setted from screen A. But i would like to dispose the state immediately after i leave screen B and not delay it.

**Comment by rrousselGit** [comment]
> Maybe navigate screen takes a microsecond or couple seconds

Navigating to a screen is not duration based. It's event-loop based. It'll take one cycle of the event loop 

As such, any non-zero duration will do. 

Also, "disposing immediately" doesn't matter. What's wrong with delaying the dispose by a few seconds?

**Comment by rrousselGit** [comment]
> One more thing, the fact that this feature doesn't solve ALL scenarios is irrrelavnt since it obsivously helps solve some common scenarios.

It is important, since there are workarounds.
Adding a new API comes with a significant cost



> How exactly can we achieve this behaviour ourselves with the current implementation?

You can do the `ref.listen` variant I showed by doing:

```dart
final container = ProviderScope.containerOf(context);
final sub = container.listen(...);
await push();
sub.close();
```

This works today.

And I have an streamlining this a bit by adding a `WidgetRef.listen` variant which returns the subscription instead of being used inside `build` like the current one.

**Comment by danielkerwin** [comment]
I agree with others that a temporary listener, or time based condition, feel awkward - what if the detailed components that subscribe to the provider themselves depend on some other unknown time-based async event, like fetching data before subscribing; the delay can't be set programmatically, but a dispose after first listen can.

**Comment by rrousselGit** [comment]
> I agree with others that a temporary listener, or time based condition, feel awkward - what if the detailed components that subscribe to the provider themselves depend on some other unknown time-based async event, like fetching data before subscribing; the delay can't be set programmatically, but a dispose after first listen can.

You can close the temporary listener whenever you want.

Same goes for the initialization. You'll call `ref.read` when you want. If your onTap has other async functions, they'd likely execute _before_ the `ref.read` – since the read will depend on them.

**Comment by talent-apps** [comment]
@rrousselGit 
The entire point of this feature is separation of concerns.
Creating an ad-hoc listenter where it is not needed seems a bit weird.

**Comment by rrousselGit** [comment]
There's no reason for it to feel weird. It's a fairly common pattern for testing

```dart
final container = ProviderContainer();
final sub = container.listen(autoDisposeProvider, (prev, value) {});

expect(sub.read(), 42);
```

That's the expected way to "read" autoDispose providers.

**Comment by talent-apps** [comment]
Just for the sake of the example, if the navigation to the screen that uses that auto disposable provider is done in 10 different places, each one would have to be aware of this and create its own dummy listener, instead of encapsulating this behaviour in the provider itself.

**Comment by rrousselGit** [comment]
What I really want to avoid is adding an option flag.  
Options tend to increase the complexity of the API quite a bit.


If this behavior is really desirable, I would prefer having a breaking change and making this the new behavior for all autoDispose providers.
Having such behavior as a flag IMO is a bad idea


It's actually something I originally wanted to do for the 1.0.0 (cf https://github.com/rrousselGit/river_pod/issues/531).

**Comment by robpot95** [comment]
I agree with option flag it can get narly, but i feel like it's described it should be the default behaviour of autoDispose

**Comment by XuanTung95** [comment]
> Example:
> 
> Some function on screen A initalizes a provider with some initial value and then navigates to screen B.
> Only screen B watches the provider.
> Once screen B is disposed, the provider should be disposed.

@talent-apps 
I wonder if this solution solves your problems?
- You read on screen A without `.autoDispose`
- When screen B is disposed, you call `invalidate()` to dispose the provider manually.

**Comment by talent-apps** [comment]
@XuanTung95 
Interesting idea, it's worth testing it.

**Comment by rrousselGit** [comment]
Might as well do what I suggested before

https://github.com/rrousselGit/river_pod/issues/1329#issuecomment-1085868131

**Comment by robpot95** [comment]
> ```dart
> final sub = ref.listen(provider, (prev, value) {});
> ```

I've tested it out and seems not to work. The value is disposed when i switch to screen2 but is accessible when i go to screen3

```
final counterProvider = StateProvider.autoDispose<int>((ref) {
  ref.keepAlive();
  ref.onCancel(ref.invalidateSelf);
  return 0;
});

class Screen1 extends ConsumerWidget {
  const Screen1({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Center(
      child: ElevatedButton(
        child: const Text("Go to screen2"),
        onPressed: () async {
          ref.read(counterProvider.state).state = 10;
          await Navigator.push(context, MaterialPageRoute(builder: (context) => const Screen2()));
        },
      ),
    );
  }
}

class Screen2 extends ConsumerWidget {
  const Screen2({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final counter = ref.watch(counterProvider);
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Text(counter.toString()),
          ElevatedButton(
            child: const Text("Go to screen3"),
            onPressed: () {
              ref.read(counterProvider.state).state = 10;
              Navigator.push(context, MaterialPageRoute(builder: (context) => const Screen3()));
            },
          ),
        ],
      ),
    );
  }
}

class Screen3 extends ConsumerWidget {
  const Screen3({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final counter = ref.watch(counterProvider);
    return Center(
      child: Text(counter.toString()),
    );
  }
}
```

**Comment by rrousselGit** [comment]
Well that's doing exactly what this issue is suggesting

This feature request would have the same problem.

**Comment by rrousselGit** [comment]
But to be fair, it sounds like you folks want `cacheTime` instead of `disposeDelay`

**Comment by robpot95** [comment]
ref.read should maintain the state of auto-dispose providers. I dont see why it gets disposed when clearly listening to the provider on next screen. CacheTime is great, but I dont think it's the solution to this problem.

Setting a time may cause it to dispose to early or keep the state for to long.

**Comment by rrousselGit** [comment]
cacheTime is the expected solution. Destroying the value as soon as the user leaves the screen is rarely the expected UX.  
Users typically would expect that quickly leaving and coming back to a screen bypasses the loading state.

If anything, cacheTime should default to a few minutes. That's how it is in other techs

**Comment by XuanTung95** [comment]
> But to be fair, it sounds like you folks want cacheTime instead of disposeDelay

@rrousselGit 

- I don't think people will like magic number in the code.
- I think the problem is `.autoDispose` does not work as expected in some special cases. Changing the behavior of `.autoDispose` might cause other problems.
- I'm thinking about giving more control to dispose a provider. Like, turn on/off autoDispose at run time or smth. So if provider is disposed at the wrong place, just tell people to turn on/off autoDispose at the right time. What do u think?

**Comment by robpot95** [comment]
> cacheTime is the expected solution. Destroying the value as soon as the user leaves the screen is rarely the expected UX. Users typically would expect that quickly leaving and coming back to a screen bypasses the loading state.
> 
> If anything, cacheTime should default to a few minutes. That's how it is in other techs

Then you are stuck with a state for to long. E.g if you fetch news, switch screen and go back. I would like to refetch the news. By dispose the old one as soon i left. In this case i dont like to have magic numbers.

**Comment by rrousselGit** [comment]
> Then you are stuck with a state for to long. E.g if you fetch news, switch screen and go back. I would like to refetch the news. By dispose the old one as soon i left. In this case i dont like to have magic numbers.

That's only temporary. There's a separate feature in progress that'll separate "call dispose" from "keep the value"


> I'm thinking about giving more control to dispose a provider. Like, turn on/off autoDispose at run time or smth. So if provider is disposed at the wrong place, just tell people to turn on/off autoDispose at the right time. What do u think?

You already have all the control you need

there's ref.keepAlive/onCancel/onResume/invalidateSelf and cacheTime+disposeDelay, and also the the typical "listen + close subscription"

**Comment by XuanTung95** [comment]
> You already have all the control you need
> 
> there's ref.keepAlive/onCancel/onResume/invalidateSelf and cacheTime+disposeDelay, and also the the typical "listen + close subscription"

@rrousselGit 

OK I think that will work.
Just some comment.

- keepAlive is in AutoDisposeRef so I need to store `ref` for every provider in order to call `keepAlive`. Could you make it easier to use?
- If I have a normal provider or family which is not `.autoDispose`. Later I want to `.autoDispose` it, could I have a method `dontKeepAlive`? The workaround would be to create `.autoDispose` and set `keepAlive` then store the KeepAliveLink, then close() it, which is a lot of code.

**Comment by rrousselGit** [comment]
> * keepAlive is in AutoDisposeRef so I need to store `ref` for every provider in order to call `keepAlive`. Could you make it easier to use?

keepAlive is private to the provider.

If you want to maintain the state of a provider from outside the provider, use `listen`+`close`

> * If I have a normal provider or family which is not `.autoDispose`. Later I want to `.autoDispose` it, could I have a method `dontKeepAlive`? The workaround would be to create `.autoDispose` and set `keepAlive` then store the KeepAliveLink, then close() it, which is a lot of code.

I don't understand what you're trying to do. Could you share some code?

**Comment by talent-apps** [comment]
@rrousselGit 
Re: "Destroying the value as soon as the user leaves the screen is rarely the expected UX.
Users typically would expect that quickly leaving and coming back to a screen bypasses the loading state."

Not sure why you assume this. The scenario that was presented earlier (list of items and a screen to edit an existing item) requires initialization of the provider every time rather than caching a previous value.

**Comment by rrousselGit** [comment]
> Not sure why you assume this. The scenario that was presented earlier (list of items and a screen to edit an existing item) requires initialization of the provider every time rather than caching a previous value.

Re-initializing the provider when re-opening the screen and keeping the previous state are two different things

They aren't mutually exclusive, and you can do both at the same time. As I mentioned, there's a wip feature for this.

**Comment by talent-apps** [comment]
@rrousselGit 
The suggested solution of https://github.com/rrousselGit/river_pod/issues/1329#issuecomment-1085868131 doesn't seem to work. How should we implement it with the current lib?

**Comment by rrousselGit** [comment]
It should work. What do you mean by "doesn't work"?

**Comment by talent-apps** [comment]
Check out https://github.com/rrousselGit/river_pod/issues/1329#issuecomment-1086828636
"The value is disposed when i switch to screen2"

**Comment by XuanTung95** [comment]
> I don't understand what you're trying to do. Could you share some code?

Example:
```
class MyState extends ChangeNotifier {
}

final myProvider = ChangeNotifierProvider.family((ref, String id) {
  return MyState();
});

class MyWidget extends ConsumerStatefulWidget {
  const MyWidget({Key? key}) : super(key: key);

  @override
  _MyState createState() => _MyState();
}

class _MyState extends ConsumerState<MyWidget> {
  final String myId = "id";
  
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    // ref.invalidate(myProvider(myId)); // <-- this will cause the provider to be disposed immediately.
    
    ref.autoDispose(myProvider(myId)); // <-- this will cause the provider to be disposed if there is no listener.
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final state = ref.watch(myProvider(myId));
    return Container();
  }
}
```

**Comment by rrousselGit** [comment]
That looks like a bug with StateProvider (& likely State/ChangeNotifierProvider)

`Provider` doesn't behave this way. Will fix

**Comment by XuanTung95** [comment]
> Well ultimately you can do that yourself already
> 
> Provider.autoDispose((ref) {
>   ref.keepAlive();
>   ref.onCancel(ref.invalidateSelf);
> }
> This effectively produces the same result as autoDispose, without the state getting destroyed during navigation

> Check out https://github.com/rrousselGit/river_pod/issues/1329#issuecomment-1086828636
> That looks like a bug with StateProvider (& likely State/ChangeNotifierProvider)

@rrousselGit 

I'm confused. 

- `ref.keepAlive()` : Requests for the state of a provider to not be disposed when all the listeners of the provider are removed.
- `ref.onCancel(ref.invalidateSelf)`:  when the last listener of the provider is removed, invalidates the state of the provider.

With this code, I expected `ref.onCancel(ref.invalidateSelf)` will cause invalidates the state of the provider before go to screen 2 because `ref.invalidateSelf` is called. If the expected behavior is keeping the state during navigation then no one will understand why, it's more like a bug to me if you implement that.

**Comment by rrousselGit** [comment]
> With this code, I expected ref.onCancel(ref.invalidateSelf) will cause invalidates the state of the provider before go to screen 2 because ref.invalidateSelf is called.

No it won't.

`ref.read` does not triggers `onCancel` 

onCancel triggers when the last listener is removed. `ref.read` doesn't add/remove listeners, so it won' trigger onCancel,


The bug with StateProvider is because it's implemented as two providers (StateProvider vs StateProvider.notifier, where the former "watch" the later). So with autoDispose, a listener is added then removed immediately

**Comment by rrousselGit** [comment]
But I think I'll make a breaking change and:

- prevent using ProviderContainer.read on autoDispose providers
- make WidgerRef.read/Ref.read "listen" to the provider indefinitely (until the associated widget/provider is disposed that is)

So doing:

```dart
Button(
  onTap: () => ref.read(autoDisposeProvider.notifier).state++,
)
```

and then tapping the button will prevent the provider from getting disposed until the button is unmounted.

**Comment by TimWhiting** [comment]
I like this idea of the breaking change. I have definitely run into problems related to those apis.

**Comment by XuanTung95** [comment]
> Although I guess this won't cause the provider to be disposed when leaving screen2 and we'd have to leave screen 1. 🤔

@rrousselGit 
Let's consider it a bit.

I think people still want to read a provider everywhere without the need to keep it alive. If everything is "watch" then it's hard to auto dispose a provider because we need to care about every read in the whole app.

The scenario of this ticket can be solved easily by using `ref.watch` on screen 1, it is just folks don't want rebuild when it's not necessary.

I think we should work on how to "read" an autoDispose provider and keep it alive until it's done.

**Comment by talent-apps** [comment]
@XuanTung95
I agree that making the `read` API into a `watch` API is counter intuitive.
The scenatio that was initially presented was that screen 1 doesn't care about the provider, and therefore doesn't need to watch it, it only initialized it with some value for screen 2.

**Comment by XuanTung95** [comment]
> If I have a normal provider or family which is not .autoDispose. Later I want to .autoDispose it, could I have a method dontKeepAlive? The workaround would be to create .autoDispose and set keepAlive then store the KeepAliveLink, then close() it, which is a lot of code.

> I don't understand what you're trying to do. Could you share some code?

@rrousselGit 

I edit my example. It looks like this.

```
final myProvider = ChangeNotifierProvider((ref) {
  Future.delayed(Duration(seconds: 2)).then((value) {
    ref.autoDispose(); // <- this provider become autoDispose from now on
  });
  Future.delayed(Duration(seconds: 5)).then((value) {
    ref.keepAlive(); // <- back to normal
  });
  return ChangeNotifier();
});
```

Goal: Add 2 method `.autoDispose()` and `.keepAlive()` to `Ref<State extends Object?>`

- `.keepAlive()`: works same as normal `.keepAlive()` of AutoDisposeRef. Maybe it return void instead of a Link object, because we can use `.autoDispose()` to change the behavior.
- `.autoDispose()`: after call this method, provider work same as normal .autoDispose provider until .keepAlive() is called.

I think this feature will complete the `.keepAlive()` implementation you did. Allow folks to controller the life-cycle of provider more flexible.
It can solve the issue of this ticket by using `.autoDispose()` method once screen B is disposed.

**Comment by XuanTung95** [comment]
I know, but it's only possible for .autoDispose, other providers also need to be disposed if nessesarry. The current implementation is not friendly if I don't want to use .autoDispose.

**Comment by rrousselGit** [comment]
That's not a bug, that's a feature. Marking a provider as .autoDispose is critical for making hard-to-catch bugs a compilation error instead. It being necessary is voluntary and this will not be changed.

**Comment by XuanTung95** [comment]
> .autoDispose is critical for making hard-to-catch bugs a compilation error instead

Could you give an example bug?

**Comment by XuanTung95** [comment]
> This is currently voluntarily made to be a compilation error

I understand.
But it has nothing to do with my proposal. The normal provider still cannot watch an autoDispose provider. It just add a away to dispose the normal provider.

**Comment by XuanTung95** [comment]
- ref.autoDispose() mean I want to dispose it if there is no listener
- ref.invalidateSelf() mean I want to dispose it even if there are listeners (if i don't misunderstand about this method)

They are 2 different use cases. If I call ref.autoDispose(), I should expect it do not dispose if someone is listening.
So how can it be a bug if I want it to behave that way.

**Comment by rrousselGit** [comment]
On a different note, what do you think about a simplification of `ref.listen` + `close`


```dart
onTap: () {
  await ref.run(provider, (value) async {
    // provider will stay alive until the callback completes
    await Navigator.push(); // The await here will keep the provider alive until the route is poped
  });
}
```

**Comment by robpot95** [comment]
> On a different note, what do you think about a simplification of `ref.listen` + `close`
> 
> ```dart
> onTap: () {
>   await ref.run(provider, (value) async {
>     // provider will stay alive until the callback completes
>     await Navigator.push(); // The await here will keep the provider alive until the route is poped
>   });
> }
> ```

This make more sense, LGTM

**Comment by XuanTung95** [comment]
> On a different note, what do you think about a simplification of `ref.listen` + `close`
> 
> ```dart
> onTap: () {
>   await ref.run(provider, (value) async {
>     // provider will stay alive until the callback completes
>     await Navigator.push(); // The await here will keep the provider alive until the route is poped
>   });
> }
> ```

I think we can do the same with this. 

```
onTap: () {
  final value = ref.read(provider);
  final link = value.keepAlive();
  await Navigator.push();
  link.close();
}
```

Adding run() method may be not necessary unless it is used a lot.

**Comment by rrousselGit** [comment]
You cannot call `keepAlive` like you've described.  
And you also need to handle exceptions, as if somehow code after the `read` throws, the provider will never get disposed.

> And "go_router" package does not return Future when navigating, so it only work for Navigation 1.0.

That's not Riverpod's fault though. You can make a feature request for it in go_router's repository

**Comment by XuanTung95** [comment]
It has been a while, any update on this ticket?

**Comment by rrousselGit** [comment]
I'm still skeptical about this. I think this would be a hazardous thing to implement and could cause quite a bit of confusion
I'd prefer looking at a better solution. As once published, the harm done by such a feature could be enormous.


For this issue, I think a better alternative would be to give you the primitives to be able to implement this on your own.
I think the `ref.onCancel`/`ref.onResume` life-cycles are good candidates (combined with `ref.keepAlive()`)

You should be able to use those to implement this on your own, without needing Riverpod to make this an official API.

**Comment by rrousselGit** [comment]
Closing in favor of https://github.com/rrousselGit/riverpod/issues/1665. That issue would add the missing feature for implementing what's described here.

**Comment by AhmedLSayed9** [comment]
@rrousselGit 
Should we reopen this?

It was closed in favor of `ref.onCancel` but ref.onCancel is not reliable to be used for this use-case as described at #2992 & #3759.

**Comment by rrousselGit** [comment]
Nope. I have no plan to change this.

IMO this feature is too dangerous, as if the "first listener" is never added for some reason, the disposal will never happen.

We can raise new issues with regads to documenting how to init some values before pushing a route though.
